Modules/_asynciomodule.c

```diff
@@ -597,12 +597,27 @@ future_set_exception(asyncio_state *state, FutureObj *fut, PyObject *exc)
         PyErr_SetString(PyExc_TypeError, "invalid exception object");
         return NULL;
     }
-    if (Py_IS_TYPE(exc_val, (PyTypeObject *)PyExc_StopIteration)) {
+    if (PyObject_IsInstance(exc_val, (PyObject *)PyExc_StopIteration)) {
```

Comment:
Do not use `PyObject_IsInstance()`. Even if it is closer to Python code, it is not used in `except`, and it can fail, returning -1.

Better use `PyErr_GivenExceptionMatches()`.